// langsmith/dist/experimental/vercel/telemetry.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LangSmithTelemetry = LangSmithTelemetry;
const run_trees_js_1 = require("../../run_trees.cjs");
const traceable_js_1 = require("../../traceable.cjs");
const env_js_1 = require("../../env.cjs");
const utils_js_1 = require("./utils.cjs");
const utils_js_2 = require("./utils.cjs");
const types_js_1 = require("../../utils/types.cjs");
function _formatMessages(messages) {
    if (!Array.isArray(messages))
        return messages;
    return messages.map((msg) => (0, utils_js_1.convertMessageToTracedFormat)(msg));
}
// oxlint-disable-next-line typescript/no-explicit-any
function _formatToolCalls(toolCalls) {
    return toolCalls.map((tc) => ({
        id: tc.toolCallId,
        type: "function",
        function: {
            name: tc.toolName,
            arguments: (0, types_js_1.isPrimitive)(tc.input)
                ? String(tc.input)
                : JSON.stringify(tc.input),
        },
    }));
}
function _formatStepOutput(
// oxlint-disable-next-line typescript/no-explicit-any
event, traceRawHttp) {
    // Build an assistant-style message from the step result
    const output = { role: "assistant" };
    // Text content
    if (event.content != null) {
        output.content = event.content;
    }
    else if (event.text != null) {
        output.content = event.text;
    }
    // Tool calls
    if (Array.isArray(event.toolCalls) && event.toolCalls.length > 0) {
        output.tool_calls = _formatToolCalls(event.toolCalls);
    }
    if (event.finishReason != null) {
        output.finish_reason = event.finishReason;
    }
    if (traceRawHttp) {
        if (event.request != null)
            output.request = event.request;
        if (event.response != null)
            output.response = event.response;
    }
    return (0, utils_js_1.convertMessageToTracedFormat)(output);
}
function _getLsAgentType(parentRunTree) {
    if ((0, run_trees_js_1.isRunTree)(parentRunTree) && parentRunTree.run_type === "tool") {
        return "subagent";
    }
    return "root";
}
function _hasNonzeroUsageMetadata(runTree) {
    const usageMetadata = runTree.extra?.metadata?.usage_metadata;
    if (!(0, types_js_1.isRecord)(usageMetadata))
        return false;
    return ["input_tokens", "output_tokens", "total_tokens"].some((key) => typeof usageMetadata[key] === "number" && usageMetadata[key] > 0);
}
/**
 * Creates a LangSmith `Telemetry` for the Vercel AI SDK.
 *
 * This adapter implements the Vercel AI SDK's `Telemetry` interface
 * and maps lifecycle events to LangSmith traces. It creates a root span for
 * the entire generation, child LLM spans for each step, and tool spans for
 * tool calls.
 *
 * ```ts
 * import { generateText, registerTelemetry } from "ai";
 * import { LangSmithTelemetry } from "langsmith/experimental/vercel";
 *
 * registerTelemetry(LangSmithTelemetry());
 *
 * const result = await generateText({
 *   model: openai("gpt-4o"),
 *   prompt: "Hello!",
 * });
 * ```
 *
 * @experimental Only available in Vercel AI SDK 7.
 */
function LangSmithTelemetry(config) {
    const { name: customName, runType = "chain", metadata: customMetadata, tags: customTags, client, projectName, processInputs, processOutputs, processChildLLMRunInputs, processChildLLMRunOutputs, traceResponseMetadata, traceRawHttp, tracingEnabled, extra: customExtra, } = config ?? {};
    function formatToolResultMessage(toolCallId, toolName, toolOutput) {
        const rawOutput = (0, types_js_1.isRecord)(toolOutput)
            ? toolOutput.type === "tool-result"
                ? toolOutput.output
                : "error" in toolOutput
                    ? toolOutput.error
                    : toolOutput
            : toolOutput;
        const artifact = rawOutput instanceof Error ? rawOutput.message : rawOutput;
        const content = typeof artifact === "string"
            ? artifact
            : (JSON.stringify(artifact) ?? String(artifact));
        return {
            role: "tool",
            content,
            tool_call_id: toolCallId,
            name: toolName,
            artifact,
        };
    }
    function appendToolResultMessage(harnessState, toolCallId, toolName, toolOutput) {
        const message = formatToolResultMessage(toolCallId, toolName, toolOutput);
        harnessState.completedStepMessages.push(message);
        return message;
    }
    function getOpenStepOrRoot(state) {
        let openStep;
        state.stepRunTrees.forEach((stepRt) => {
            if (stepRt.end_time == null) {
                openStep = stepRt;
            }
        });
        return openStep ?? state.rootRunTree;
    }
    async function flushPiDeferredStepRun(state) {
        const workaround = state.experimental_harness?.piAggregateUsageWorkaround;
        if (workaround == null)
            return;
        const stepRunTree = workaround.deferredStepRunTree;
        if (stepRunTree == null)
            return;
        workaround.deferredStepRunTree = undefined;
        await stepRunTree.patchRun({ excludeInputs: true });
    }
    async function finalizeOpenToolRuns(state, opts) {
        const entries = Array.from(state.toolRunTrees.entries());
        for (let i = 0; i < entries.length; i++) {
            const [, toolRt] = entries[i];
            if (toolRt.end_time == null) {
                if (opts?.error != null) {
                    await toolRt.end(undefined, opts.error);
                }
                else {
                    await toolRt.end(opts?.note != null ? { note: opts.note } : undefined);
                }
                await toolRt.patchRun({ excludeInputs: true });
            }
        }
        state.toolRunTrees.clear();
    }
    /** Per-generation state keyed by AI SDK `callId` (stable across nested calls). */
    const invocationsByCallId = new Map();
    const onStart = async (event) => {
        if (!(0, env_js_1.isEnvTracingEnabled)(tracingEnabled))
            return;
        if (!("callId" in event) || typeof event.callId !== "string")
            return;
        // If called within an existing traceable context, nest under it
        const parentRunTree = (0, traceable_js_1.getCurrentRunTree)(true);
        let inputs = {};
        if (event.recordInputs !== false) {
            if ("messages" in event && event.messages != null) {
                inputs.messages = _formatMessages(event.messages);
            }
            if ("prompt" in event && event.prompt != null) {
                inputs.prompt = event.prompt;
            }
            if ("instructions" in event && event.instructions != null) {
                inputs.instructions = event.instructions;
            }
            if ("system" in event && event.system != null) {
                inputs.system = event.system;
            }
            if ("tools" in event && event.tools != null) {
                inputs.tools = Object.keys(event.tools);
            }
            if ("runtimeContext" in event && event.runtimeContext != null) {
                inputs.runtimeContext = event.runtimeContext;
            }
            if ("toolsContext" in event && event.toolsContext != null) {
                inputs.toolsContext = event.toolsContext;
            }
            // Apply user-provided input processing
            if (processInputs) {
                try {
                    inputs = processInputs(inputs);
                }
                catch (e) {
                    console.error("Error in processInputs, using raw inputs:", e);
                }
            }
        }
        const runTreeConfig = {
            name: customName ?? event.functionId ?? event.provider,
            run_type: runType,
            inputs,
            tracingEnabled: true,
            extra: {
                ...customExtra,
                metadata: {
                    ...customMetadata,
                    ai_sdk_method: event.operationId,
                    ls_agent_type: _getLsAgentType(parentRunTree),
                    ls_model_name: event.modelId,
                    ls_integration: "vercel-ai-sdk-telemetry",
                },
            },
            tags: customTags,
            ...(client ? { client } : {}),
            ...(projectName ? { project_name: projectName } : {}),
        };
        let rootRunTree;
        if ((0, run_trees_js_1.isRunTree)(parentRunTree)) {
            rootRunTree = parentRunTree.createChild(runTreeConfig);
        }
        else {
            rootRunTree = new run_trees_js_1.RunTree(runTreeConfig);
        }
        await rootRunTree.postRun();
        invocationsByCallId.set(event.callId, {
            rootRunTree,
            stepRunTrees: new Map(),
            toolRunTrees: new Map(),
            experimental_harness: event.operationId === "ai.harness"
                ? {
                    lastStepContent: [],
                    completedStepMessages: [],
                    toolResults: new Map(),
                    piAggregateUsageWorkaround: event.provider === "pi" ? {} : undefined,
                }
                : undefined,
        });
    };
    const onStepStart = async (event) => {
        const state = invocationsByCallId.get(event.callId);
        if (!state)
            return;
        await flushPiDeferredStepRun(state);
        const stepNumber = event.stepNumber ?? 0;
        let inputs = {};
        if (event.recordInputs !== false) {
            if ("messages" in event && event.messages != null) {
                inputs.messages = [
                    ..._formatMessages(event.messages),
                    ...(state.experimental_harness?.completedStepMessages ?? []),
                ];
            }
            if ("runtimeContext" in event && event.runtimeContext != null) {
                inputs.runtimeContext = event.runtimeContext;
            }
            if ("toolsContext" in event && event.toolsContext != null) {
                inputs.toolsContext = event.toolsContext;
            }
            if (processChildLLMRunInputs) {
                try {
                    inputs = processChildLLMRunInputs(inputs);
                }
                catch (e) {
                    console.error("Error in processChildLLMRunInputs, using raw inputs:", e);
                }
            }
        }
        const stepRunTree = state.rootRunTree.createChild({
            name: event.provider,
            run_type: "llm",
            inputs,
            extra: {
                metadata: {
                    step_number: stepNumber,
                    ls_model_name: event.modelId,
                },
            },
        });
        state.stepRunTrees.set(stepNumber, stepRunTree);
        const piUsageWorkaround = state.experimental_harness?.piAggregateUsageWorkaround;
        if (piUsageWorkaround != null) {
            piUsageWorkaround.latestStepRunTree = stepRunTree;
        }
        await stepRunTree.postRun();
    };
    const onLanguageModelCallStart = async (event) => {
        const state = invocationsByCallId.get(event.callId);
        if (!state)
            return;
        const stepRunTree = getOpenStepOrRoot(state);
        if (stepRunTree.run_type !== "llm")
            return;
        const prevParams = (0, types_js_1.isRecord)(stepRunTree.extra?.invocation_params)
            ? stepRunTree.extra.invocation_params
            : {};
        const nextParams = { ...event };
        // Remove properties that are already in the step run tree
        delete nextParams.messages;
        delete nextParams.provider;
        delete nextParams.modelId;
        // Remove telemetry options (except functionId)
        delete nextParams.recordInputs;
        delete nextParams.recordOutputs;
        delete nextParams.includeToolsContext;
        // Massage tools for LangSmith to render schema nicely
        nextParams.tools = nextParams.tools?.map((tool) => {
            const newTool = { ...tool };
            if ("inputSchema" in newTool) {
                newTool.input_schema = newTool.inputSchema;
                delete newTool.inputSchema;
            }
            return newTool;
        });
        stepRunTree.extra = {
            ...stepRunTree.extra,
            invocation_params: { ...prevParams, ...nextParams },
        };
    };
    const onToolExecutionStart = async (event) => {
        const state = invocationsByCallId.get(event.callId);
        if (!state)
            return;
        const parentRunTree = state.rootRunTree;
        let inputs = {};
        if (event.recordInputs !== false) {
            let toolCallInput = event.toolCall.input;
            if (typeof toolCallInput === "string") {
                try {
                    toolCallInput = JSON.parse(toolCallInput);
                }
                catch {
                    // The tool argument itself may be a plain string rather than JSON.
                }
            }
            if ((0, types_js_1.isRecord)(toolCallInput)) {
                inputs = { ...toolCallInput };
            }
            else if (toolCallInput !== undefined) {
                inputs = { input: toolCallInput };
            }
        }
        const toolRunTree = parentRunTree.createChild({
            name: event.toolCall.toolName,
            run_type: "tool",
            inputs,
            extra: {
                metadata: {
                    tool_call_id: event.toolCall.toolCallId,
                    ai_sdk_call_id: event.callId,
                },
            },
        });
        await toolRunTree.postRun();
        state.toolRunTrees.set(event.toolCall.toolCallId, toolRunTree);
    };
    const onToolExecutionEnd = async (event) => {
        const state = invocationsByCallId.get(event.callId);
        if (!state)
            return;
        const toolRunTree = state.toolRunTrees.get(event.toolCall.toolCallId);
        if (!toolRunTree)
            return;
        const harnessState = state.experimental_harness;
        if (harnessState != null) {
            harnessState.toolResults.set(event.toolCall.toolCallId, {
                toolName: event.toolCall.toolName,
                output: event.toolOutput,
            });
        }
        let outputs;
        let error;
        if (event.recordOutputs !== false) {
            if (event.toolOutput.type === "tool-result") {
                outputs = { output: event.toolOutput.output };
            }
            else if ((0, types_js_1.isRecord)(event.toolOutput) && "error" in event.toolOutput) {
                const err = event.toolOutput.error;
                error = err instanceof Error ? err.message : String(err);
            }
        }
        else {
            outputs = {};
        }
        await toolRunTree.end(outputs, error, Math.floor(toolRunTree.start_time + event.toolExecutionMs));
        await toolRunTree.patchRun({ excludeInputs: true });
        state.toolRunTrees.delete(event.toolCall.toolCallId);
    };
    const onStepFinish = async (event) => {
        const state = invocationsByCallId.get(event.callId);
        if (!state)
            return;
        const stepNumber = event.stepNumber ?? 0;
        const stepRunTree = state.stepRunTrees.get(stepNumber);
        if (!stepRunTree)
            return;
        const harnessState = state.experimental_harness;
        let outputs = {};
        if (event.recordOutputs !== false) {
            if (harnessState != null && Array.isArray(event.content)) {
                harnessState.lastStepContent = [...event.content];
            }
            outputs = _formatStepOutput(event, traceRawHttp);
            if (processChildLLMRunOutputs) {
                try {
                    outputs = processChildLLMRunOutputs(outputs);
                }
                catch (e) {
                    console.error("Error in processChildLLMRunOutputs, using raw outputs:", e);
                }
            }
        }
        if (harnessState != null && Array.isArray(event.content)) {
            harnessState.completedStepMessages.push((0, utils_js_1.convertMessageToTracedFormat)({
                role: "assistant",
                content: event.content,
            }));
            for (const part of event.content) {
                if (!(0, types_js_1.isRecord)(part) ||
                    part.type !== "tool-call" ||
                    typeof part.toolCallId !== "string") {
                    continue;
                }
                const toolResult = harnessState.toolResults.get(part.toolCallId);
                if (toolResult) {
                    appendToolResultMessage(harnessState, part.toolCallId, toolResult.toolName, toolResult.output);
                    harnessState.toolResults.delete(part.toolCallId);
                }
            }
        }
        // Set usage metadata
        // @ts-expect-error SharedV4ProviderMetadata is not assignable to SharedV2ProviderMetadata
        (0, utils_js_2.setUsageMetadataOnRunTree)(event, stepRunTree);
        await stepRunTree.end(outputs);
        const piUsageWorkaround = harnessState?.piAggregateUsageWorkaround;
        if (piUsageWorkaround != null && !_hasNonzeroUsageMetadata(stepRunTree)) {
            // Pi reports zero usage for inferred step boundaries and sends the real
            // aggregate at turn end. Defer the final PATCH so that aggregate can be
            // attached to the last LLM run without issuing a duplicate update.
            piUsageWorkaround.deferredStepRunTree = stepRunTree;
        }
        else {
            await stepRunTree.patchRun({ excludeInputs: true });
        }
        state.stepRunTrees.delete(stepNumber);
    };
    const onEnd = async (event) => {
        if (!("callId" in event) || typeof event.callId !== "string")
            return;
        const state = invocationsByCallId.get(event.callId);
        if (!state)
            return;
        const { rootRunTree } = state;
        const harnessState = state.experimental_harness;
        const piUsageWorkaround = harnessState?.piAggregateUsageWorkaround;
        await finalizeOpenToolRuns(state, { note: "closed on finish" });
        // Ensure any remaining step runs are closed
        const remainingSteps = Array.from(state.stepRunTrees.entries());
        for (let i = 0; i < remainingSteps.length; i++) {
            const [stepNumber, stepRt] = remainingSteps[i];
            if (stepRt.end_time == null) {
                await stepRt.end({ note: "closed on finish" });
                if (piUsageWorkaround != null && !_hasNonzeroUsageMetadata(stepRt)) {
                    piUsageWorkaround.deferredStepRunTree = stepRt;
                }
                else {
                    await stepRt.patchRun({ excludeInputs: true });
                }
            }
            state.stepRunTrees.delete(stepNumber);
        }
        let outputs = {};
        if (event.recordOutputs !== false) {
            // Final result output
            if ("text" in event && event.text != null) {
                outputs.content = event.text;
            }
            else if ("content" in event &&
                event.content != null &&
                (!Array.isArray(event.content) || event.content.length > 0)) {
                outputs.content = event.content;
            }
            else if (harnessState != null &&
                harnessState.lastStepContent.length > 0) {
                // HarnessAgent currently reports an empty root `content` array even
                // though each completed step contains assistant output. Only the final
                // step belongs on the root; prior steps have their own child runs.
                outputs.content = harnessState.lastStepContent;
            }
            if (outputs.content != null) {
                outputs.role = "assistant";
            }
            if ("object" in event && event.object != null) {
                outputs.object = event.object;
            }
            if ("toolCalls" in event &&
                Array.isArray(event.toolCalls) &&
                event.toolCalls.length > 0) {
                outputs.tool_calls = _formatToolCalls(event.toolCalls);
            }
            if ("finishReason" in event && event.finishReason != null) {
                outputs.finish_reason = event.finishReason;
            }
            if (traceResponseMetadata &&
                "steps" in event &&
                Array.isArray(event.steps)) {
                outputs.steps = event.steps.map((step, idx) => ({
                    step_number: idx,
                    ..._formatStepOutput(step, traceRawHttp),
                }));
            }
            if (processOutputs) {
                try {
                    outputs = processOutputs(outputs);
                }
                catch (e) {
                    console.error("Error in processOutputs, using raw outputs:", e);
                }
            }
        }
        // Pi reports zero usage on inferred step boundaries and provides the actual
        // aggregate only when the turn ends. Keep that usage on an LLM run rather
        // than incorrectly elevating it to the parent chain. The latest step is the
        // only accurate place available until Harness Pi exposes per-step usage.
        const latestStep = piUsageWorkaround?.latestStepRunTree;
        if (latestStep?.run_type === "llm" &&
            !_hasNonzeroUsageMetadata(latestStep)) {
            if ("totalUsage" in event && event.totalUsage != null) {
                (0, utils_js_2.setUsageMetadataOnRunTree)(
                // @ts-expect-error SharedV4ProviderMetadata is not assignable to SharedV2ProviderMetadata
                { usage: event.totalUsage, providerMetadata: event.providerMetadata }, latestStep);
            }
            else if ("usage" in event && event.usage != null) {
                // @ts-expect-error SharedV4ProviderMetadata is not assignable to SharedV2ProviderMetadata
                (0, utils_js_2.setUsageMetadataOnRunTree)(event, latestStep);
            }
        }
        await rootRunTree.end(outputs);
        await Promise.all([
            flushPiDeferredStepRun(state),
            rootRunTree.patchRun({ excludeInputs: true }),
        ]);
        invocationsByCallId.delete(event.callId);
    };
    const onError = async (payload) => {
        const callId = typeof payload === "object" &&
            payload !== null &&
            "callId" in payload &&
            typeof payload.callId === "string"
            ? payload.callId
            : undefined;
        const error = typeof payload === "object" && payload !== null && "error" in payload
            ? payload.error
            : payload;
        if (callId === undefined)
            return;
        const state = invocationsByCallId.get(callId);
        if (!state)
            return;
        const { rootRunTree } = state;
        const errorMsg = error instanceof Error ? error.message : String(error);
        await finalizeOpenToolRuns(state, { error: errorMsg });
        // Close any open step runs with error
        const errorSteps = Array.from(state.stepRunTrees.entries());
        for (let i = 0; i < errorSteps.length; i++) {
            const [stepNumber, stepRt] = errorSteps[i];
            if (stepRt.end_time == null) {
                await stepRt.end(undefined, errorMsg);
                await stepRt.patchRun({ excludeInputs: true });
            }
            state.stepRunTrees.delete(stepNumber);
        }
        await flushPiDeferredStepRun(state);
        await rootRunTree.end(undefined, errorMsg);
        await rootRunTree.patchRun({ excludeInputs: true });
        invocationsByCallId.delete(callId);
    };
    const executeTool = async (params) => {
        const state = invocationsByCallId.get(params.callId);
        const toolRunTree = state?.toolRunTrees.get(params.toolCallId);
        if (toolRunTree != null) {
            return (0, traceable_js_1.withRunTree)(toolRunTree, () => params.execute());
        }
        return params.execute();
    };
    return {
        onStart,
        onStepStart,
        onLanguageModelCallStart,
        onToolExecutionStart,
        onToolExecutionEnd,
        onStepFinish,
        onEnd,
        onError,
        executeTool,
    };
}
