// langsmith/dist/utils/jestlike/globals.js
import { AsyncLocalStorage } from "node:async_hooks";
import { Client } from "../../client.js";
import { getEnvironmentVariable } from "../env.js";
import { isEnvTracingEnabled } from "../../env.js";
export const DEFAULT_TEST_CLIENT = new Client();
export const testWrapperAsyncLocalStorageInstance = new AsyncLocalStorage();
export function trackingEnabled(context) {
    if (typeof context.enableTestTracking === "boolean") {
        return context.enableTestTracking;
    }
    if (getEnvironmentVariable("LANGSMITH_TEST_TRACKING") === "false") {
        return false;
    }
    return isEnvTracingEnabled();
}
export const evaluatorLogFeedbackPromises = new Set();
export function _logTestFeedback(params) {
    const { exampleId, feedback, context, runTree, client, sourceRunId } = params;
    if (trackingEnabled(context)) {
        if (exampleId === undefined) {
            throw new Error("Could not log feedback to LangSmith: missing example id. Please contact us for help.");
        }
        if (runTree === undefined) {
            throw new Error("Could not log feedback to LangSmith: missing run information. Please contact us for help.");
        }
        evaluatorLogFeedbackPromises.add((async () => {
            if (context.project == null) {
                throw new Error("Could not log feedback to LangSmith: missing project information. Please contact us for help.");
            }
            await context.syncExamplePromises?.get(exampleId);
            await client?.logEvaluationFeedback({
                evaluatorResponse: feedback,
                run: runTree,
                projectId: context.project.id,
                sourceInfo: sourceRunId !== undefined
                    ? { __run: { run_id: sourceRunId } }
                    : undefined,
            });
        })());
    }
    context.onFeedbackLogged?.(feedback);
}
